lms/models/_mixins/public_id.py

```diff
@@ -1,8 +1,14 @@
+from os import environ
+
 import sqlalchemy as sa
 from sqlalchemy.ext.hybrid import Comparator, hybrid_property
 
 from lms.models.public_id import PublicId
-from lms.models.region import Regions
+from lms.models.region import Region
+
+
+def _get_current_region():
+    return Region(code=environ["REGION_CODE"], authority=environ["H_AUTHORITY"])
```

Comment:
Unfortunately there's code here in the _models_ (!) that calls `Regions.get_region()` to get the current request's region: [in `_PublicIdComparator`](https://github.com/hypothesis/lms/blob/00fa042a6629eee9bf33317ed7b211850b3ce6d9/lms/models/_mixins/public_id.py#L20) and [in `PublicIdMixin`](https://github.com/hypothesis/lms/blob/00fa042a6629eee9bf33317ed7b211850b3ce6d9/lms/models/_mixins/public_id.py#L53). I can't simply replace these with calls to the new `RegionService.get()` because models don't (and shouldn't) have access to services or to the `request` or `settings`.

There's an unusual and quite bad thing going on here on `main` where:

1. `app.py` calls `Regions.set_region()` at app creation time:

   https://github.com/hypothesis/lms/blob/00fa042a6629eee9bf33317ed7b211850b3ce6d9/lms/app.py#L18-L20

2. `Regions.set_region()` sets a class-level attribute on the `Regions` class:

   https://github.com/hypothesis/lms/blob/00fa042a6629eee9bf33317ed7b211850b3ce6d9/lms/models/region.py#L70

3. Subsequent calls to `Regions.get_region()` return the class-level attribute that was set by `set_region()`. `get_region()` crashes if `set_region()` hasn't been called first:

   https://github.com/hypothesis/lms/blob/00fa042a6629eee9bf33317ed7b211850b3ce6d9/lms/models/region.py#L43-L56

So we've implemented our own global singleton that depends on values from the request settings (`region_code` and `h_authority`) without putting it on the `request` object. And apparently we did all this so that we can use the global singleton in an SQLAlchemy comparator and hybrid property in the models.

Yikes.

I don't yet understand _why_ the models need to do this. Perhaps it can be removed. That would be nicer.

For the time being I've implemented this `_get_current_region()` function that reads `os.environ` as a workaround. This is bad: model code shouldn't be reading `os.environ`. But it might be less bad than what's currently on `main`?

Alternatively this model code may be able to use Pyramid's [`get_current_request()` (to get the current request and get the `RegionService` instance from it) or `get_current_registry()` (to get the current registry and use it to construct a `RegionService` instance directly from the settings)](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/threadlocal.html) and then call `RegionService.get()`.

Or the model code could construct a `RegionService` directly passing in `REGION_CODE` and `H_AUTHORITY` from `os.environ` and then call `RegionService.get()`.

Reply:
If we wanted to make `_get_current_region()` more official (instead of hiding it away as a hack) I think we'd make it a classmethod on `Region`. This would be an alternative to `RegionService`:

```python
@dataclass
class Region:
    """Codes and details related to a deployment region."""

    code: str
    """Region code (e.g. ISO 3166 alpha-2 country code)."""

    authority: str
    """Associated Hypothesis authority."""

    def __post_init__(self):
        names = {"us": "Worldwide (U.S.)", "ca": "Canada"}

        #: Human readable description of the region.
        self.name: str = names[self.code]

    @classmethod
    def get(cls):
        return cls(code=environ["REGION_CODE"], authority=environ["H_AUTHORITY"])
```